composite_broker/src/broker.rs

```diff
@@ -0,0 +1,298 @@
+    // }};
+    use mqtt_sn::{self, Message, Connect, ConnAck, ReturnCode, RejectedReason, Subscribe, SubAck, 
+        Flags, Publish, PubAck, UnsubAck, Unsubscribe}; // TopicName
+
+    
+    #[derive(Debug, Clone, PartialEq)]
+    #[allow(dead_code)]
+    pub struct Client {
+        client_id: String,
+        address: String,
+        port: String
+    }
+
+    impl Client {
+        fn new (client: String, address: String, port: String) -> Client {
+            Client {client_id: client, address: address, port: port}
+        }
+    }
+
+    #[derive(Debug, Clone)]
+    #[allow(dead_code)]
+    pub struct MBroker {
+        num_packets: u16,
+        num_clients: u16,
+        num_topics: u16,
+        client_list: Vec<Client>,
+        concrete_subscriptions_list: HashMap<u16, Vec<Client>>,
+        topicname_id_pairs: HashMap<String, u16>,
+    } 
+
+    #[allow(dead_code)]
+    impl MBroker {
+
+        // make a new Broker
+        pub fn new() -> Self {
+            Self { 
+                num_packets: 2000,
+                num_clients: 0,
+                num_topics: 0,
+                client_list: Vec::new(), 
+                concrete_subscriptions_list: HashMap::new(), //DashMap::new(),
+                topicname_id_pairs: HashMap::new()
+            }
+        } // end new
+
+        // accept a connect packet
+        // param: address of the client, connect packet
+        // return: connect ack
+        pub fn accept_connect(&mut self, addr: String, _connect_packet: Connect) -> Message {
+            let addr_port: Vec<&str> = addr.split(":").collect();
+            // error checking that client isn't already connected
+            if self.find_client_by_address(addr_port[0], addr_port[1]) {
+                let packet = Message::ConnAck(ConnAck { 
+                    code: ReturnCode::Rejected(RejectedReason::Congestion)
+                });
+                return packet
+            }
+            // make a new client 
+            self.num_clients += 1;
+            let c: Client = Client::new(self.num_clients.to_string(), addr_port[0].to_string(), addr_port[1].to_string());
+
+            // add to client list
+            self.client_list.push(c);
+
+            // create connect_ack packet
+            let conn_ack = Message::ConnAck(ConnAck { 
+                code: ReturnCode::Accepted 
+            } );
+            self.num_packets +=1;
+            conn_ack
+        } // end connect
+
+
+        // accept a subscribe packet -> (Packet, Option<Vec<String>>)
+        #[allow(unused_assignments)]
+        pub fn accept_sub(&mut self, addr: String, packet: Subscribe) -> Message {
+            self.num_packets +=1;
+            let addr_port: Vec<&str> = addr.split(":").collect();
+            let mut cli: Client = Client::new(String::new(), String::new(), String::new());
+            // find the client in reference
+            for client in self.client_list.clone() {
```

Comment:
This clone is copying the entire list, I'd imagine. That's pretty heavyweight. Might not be necessary with a `&self.clientlist` instead.